compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp

```diff
@@ -0,0 +1,1364 @@
+//
+// clang-format on
+template <typename ElementwiseOpT>
+struct ReorderBroadcastInDimOpAndElementwiseOp final
+    : OpRewritePattern<ElementwiseOpT> {
+  using OpRewritePattern<ElementwiseOpT>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ElementwiseOpT op,
+                                PatternRewriter &rewriter) const override {
+    Operation *operation = op.getOperation();
+    assert(operation->getNumOperands() >= 1 && operation->getNumResults() == 1);
+
+    // Verify if all operands are from BroadcastInDimOp and its
+    // broadcast_dimensions is the same.
+    llvm::SmallVector<mlir::stablehlo::BroadcastInDimOp, 2> bcastOps;
+    for (auto operand : operation->getOperands()) {
+      if (auto bcastOp =
+              operand.getDefiningOp<mlir::stablehlo::BroadcastInDimOp>()) {
+        bcastOps.push_back(bcastOp);
+      } else {
+        return failure();
+      }
+    }
+
+    if (llvm::any_of(bcastOps,
+                     [&bcastOps](mlir::stablehlo::BroadcastInDimOp bcastOp) {
+                       return bcastOp.getBroadcastDimensions() !=
+                              bcastOps[0].getBroadcastDimensions();
+                     })) {
+      return failure();
+    }
+
+    // Verify if all operands of BroadcastInDimOp are of same type and have
+    // static shape.
+    auto bcastOperandType =
+        dyn_cast<ShapedType>(bcastOps[0].getOperand().getType());
+    llvm::SmallVector<Value, 2> bcastOperands;
+    for (auto bcastOp : bcastOps) {
+      auto bcastOperand = bcastOp.getOperand();
+      auto type = dyn_cast<ShapedType>(bcastOperand.getType());
+      if (!type || !type.hasStaticShape() || type != bcastOperandType) {
+        return failure();
+      }
+      bcastOperands.push_back(bcastOperand);
+    }
+
+    // Some elementwise ops, mlir::stablehlo::RealOp for example, do not have
+    // SameOperandsAndResultType trait, so resultType might be different
+    // from bcastOperandType.
+    auto elementType = getElementTypeOrSelf(op.getResult());
+    auto resultShape = bcastOperandType.getShape();
+    auto resultType = RankedTensorType::get(resultShape, elementType);
+
+    Value result =
+        rewriter.create<ElementwiseOpT>(op.getLoc(), resultType, bcastOperands);
+    rewriter.replaceOpWithNewOp<mlir::stablehlo::BroadcastInDimOp>(
+        op, op.getType(), result, bcastOps[0].getBroadcastDimensions());
+
+    for (auto bcastOp : bcastOps) {
+      if (bcastOp.getOperation()->use_empty()) {
+        rewriter.eraseOp(bcastOp);
+      }
+    }
+
+    return success();
+  }
+};
+
+// Identifies cases where a dense operation has inputs that come from widening
+// operations. For instance, a dot product widening from FP16 to FP32 is better
+// to have the casting operation fused into the dot operation. This decreases
+// the loading required during a dense computation.
+template <class Op>
+struct FuseWidenOperands final : OpRewritePattern<Op> {
+  using OpRewritePattern<Op>::OpRewritePattern;
```

Comment:
This seems like a low level optimization. Is doing this at the input processing stage always useful?

Reply:
I generally try to avoid changing the logic while porting to keep the process simple. I'll add a todo to re-consider this fold though.